### PR TITLE
IDEFICS: allow interpolation of vision's pos embeddings

### PR DESCRIPTION
**Pull request by leot13**
# What does this PR do?

Allows vision position embeddings to be interpolated. Thus allowing bigger images to be passed to the model.
Fixes issue #26154

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?
@ArthurZucker @amyeroberts

### CODE REVIEW

File: src/transformers/models/idefics/vision.py [hunk 1]
```diff
@@ -87,15 +84,63 @@ def __init__(self, config: IdeficsVisionConfig):
...
+        patch_pos_embed = patch_pos_embed.permute(0, 3, 1, 2)
+        patch_pos_embed = nn.functional.interpolate(
+            patch_pos_embed,
+            scale_factor=(h0 / math.sqrt(num_positions), w0 / math.sqrt(num_positions)),
+            mode="bicubic",
+            align_corners=False,
+        )
+        assert int(h0) == patch_pos_embed.shape[-2] and int(w0) == patch_pos_embed.shape[-1]
+        patch_pos_embed = patch_pos_embed.permute(0, 2, 3, 1).view(1, -1, dim)
+        return torch.cat((class_pos_embed.unsqueeze(0), patch_pos_embed), dim=1)
```
Comment (amyeroberts) [inline src/transformers/models/idefics/vision.py:119]: For clearer var names 
* avoid clash with 'dim'
* h0, w0 aren't really height and width
Only do sqrt calculation once

```suggestion
        embed_dim = embeddings.shape[-1]
        num_h_patches = height // self.config.patch_size
        num_w_patches = width // self.config.patch_size
        # we add a small number to avoid floating point error in the interpolation
        # see discussion at https://github.com/facebookresearch/dino/issues/8
        num_h_patches, num_w_patches = num_h_patches + 0.1, num_w_patches + 0.1
        sqrt_num_positions = math.sqrt(num_positions)
        patch_pos_embed = patch_pos_embed.reshape(1, int(sqrt_num_positions), int(sqrt_num_positions), embed_dim)
        patch_pos_embed = patch_pos_embed.permute(0, 3, 1, 2)
        patch_pos_embed = nn.functional.interpolate(
            patch_pos_embed,
            scale_factor=(num_h_patches / sqrt_num_positions, num_w_patches / sqrt_num_positions),
            mode="bicubic",
            align_corners=False,
        )
        if (
            int(num_h_patches) != patch_pos_embed.shape[-2] 
            or int(num_w_patches) == patch_pos_embed.shape[-1]
        ):
            raise ValueError(
                f"Number of patches for images ({int(num_h_patches), int(num_w_patches)}) don't match the "
                f"shape of position embedding ({patch_pos_embed.shape[-2], patch_pos_embed.shape[-1]})"
            )
        patch_pos_embed = patch_pos_embed.permute(0, 2, 3, 1).view(1, -1, embed_dim)
        return torch.cat((class_pos_embed.unsqueeze(0), patch_pos_embed), dim=1)
```

File: src/transformers/models/idefics/vision.py [hunk 2]
```diff
@@ -380,6 +425,7 @@ class IdeficsVisionTransformer(nn.Module):
     def __init__(self, config: IdeficsVisionConfig):
         super().__init__()
         self.config = config
+        self.interpolate_pos_encoding = config.interpolate_pos_encoding
```
Comment (amyeroberts) [inline src/transformers/models/idefics/vision.py:428]: To align with other models, this should be a kwarg in the forward method, rather than set by the config

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding this! 

Few small comments - mostly structure and styling suggestions to be in line with other models

File: tests/models/idefics/test_modeling_idefics.py [hunk 3]
```diff
@@ -169,6 +169,33 @@ def prepare_config_and_inputs(self):
 
         return (config, input_ids, input_mask, pixel_values, image_attention_mask)
 
+    def prepare_config_and_inputs_for_image_pos_embeddings_interpolation(self):
+        self.seq_length = 42
```
[the remark this answers is not in the document]
Comment (leot13) [reply tests/models/idefics/test_modeling_idefics.py:173]: I reused the prepare_config_and_inputs() method to make the test for interpolation, but I am not sure why the seq_len should be changed. @stas00 probably has a better idea as to why it is here

File: src/transformers/models/idefics/vision.py (hunk 2 above)
Comment (leot13) [reply src/transformers/models/idefics/vision.py]: Indeed that makes more sense

File: tests/models/idefics/test_modeling_idefics.py [hunk 4]
```diff
@@ -169,6 +169,33 @@ def prepare_config_and_inputs(self):
...
+
+        input_ids = ids_tensor([self.batch_size, self.seq_length], self.vocab_size)
+
+        num_images = 2 if self.add_multiple_images else 1
```
[the remark this answers is not in the document]
Comment (leot13) [reply tests/models/idefics/test_modeling_idefics.py]: I fixed the tests according to your comments. Mainly, I took off the self.add_multiple_images / self.num_images, took of the change of self.seq_len, and made 3 different tests instead of 2 previously.

File: tests/models/idefics/test_modeling_idefics.py [hunk 5]
```diff
@@ -290,7 +303,8 @@ def test_training(self):
             model.train()
             inputs = self._prepare_for_class(inputs_dict, model_class, return_labels=True)
             for k, v in inputs.items():
-                print(k, v.shape)
+                if isinstance(v, torch.Tensor):
+                    print(k, v.shape)
```
Comment (amyeroberts) [inline tests/models/idefics/test_modeling_idefics.py:307]: We shouldn't have print statements in the tests

```suggestion
```

File: tests/models/idefics/test_modeling_idefics.py [hunk 6]
```diff
@@ -150,14 +146,18 @@ def __init__(
...
         self.expected_seq_len = self.seq_length + (self.image_size // self.patch_size) ** 2 + 1
 
-    def prepare_config_and_inputs(self):
-        self.seq_length = 42
-
+    def prepare_config_and_inputs(self, num_images=1, interpolate_pos_encoding=False):
         input_ids = ids_tensor([self.batch_size, self.seq_length], self.vocab_size)
 
-        num_images = 2 if self.add_multiple_images else 1
+        expand_image_size = 0 if not interpolate_pos_encoding else 2
```
Comment (amyeroberts) [inline tests/models/idefics/test_modeling_idefics.py:152]: nit: this is just an opinion, not a hard rule or one you have to implement. It's generally easier and quicker to interpret code when you're going from the positive statement that the negation 

```suggestion
        expand_image_size = 2 if interpolate_pos_encoding else 0
```

File: tests/models/idefics/test_modeling_idefics.py [hunk 7]
```diff
@@ -268,8 +273,16 @@ def setUp(self):
...
+    def test_model_single_image(self):
+        config_and_inputs = self.model_tester.prepare_config_and_inputs(num_images=1, interpolate_pos_encoding=False)
+        self.model_tester.create_and_check_model(*config_and_inputs)
+
+    def test_model_multiple_images(self):
+        config_and_inputs = self.model_tester.prepare_config_and_inputs(num_images=2, interpolate_pos_encoding=False)
+        self.model_tester.create_and_check_model(*config_and_inputs)
+
+    def test_model_with_image_pos_embeddings_interpolation(self):
+        config_and_inputs = self.model_tester.prepare_config_and_inputs(num_images=1, interpolate_pos_encoding=False)
```
Comment (amyeroberts) [inline tests/models/idefics/test_modeling_idefics.py:285]:
```suggestion
        config_and_inputs = self.model_tester.prepare_config_and_inputs(num_images=1, interpolate_pos_encoding=True)
```

File: src/transformers/models/idefics/configuration_idefics.py [hunk 8]
```diff
@@ -71,6 +71,8 @@ class IdeficsVisionConfig(PretrainedConfig):
             testing).
         initializer_range (`float`, *optional*, defaults to 0.02):
             The standard deviation of the truncated_normal_initializer for initializing all weight matrices.
+        interpolate_pos_encoding (`bool`, *optional*, defaults to `False`):
+            Activates interpolation of position embeddings, allowing for images of different sizes.
```
Comment (amyeroberts) [inline src/transformers/models/idefics/configuration_idefics.py:75]:
```suggestion
```

**Review by amyeroberts — COMMENTED** [review]
Thanks for iterating on this! 

Just some small tidy-ups: removing param name in config, removing print statements, correct setting for a test.

File: tests/models/idefics/test_modeling_idefics.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (stas00) [reply tests/models/idefics/test_modeling_idefics.py]: may be move the setting of it to:

https://github.com/huggingface/transformers/blob/857b45ce057cccad2b5c07ab41e963203c4950a6/tests/models/idefics/test_modeling_idefics.py#L54

so it's the same everywhere unless it's not

**Review by amyeroberts — APPROVED** [review]
Thanks for adding!

File: src/transformers/models/idefics/vision.py [hunk 9]
```diff
@@ -87,15 +85,79 @@ def __init__(self, config: IdeficsVisionConfig):
...
+        # see discussion at https://github.com/facebookresearch/dino/issues/8
+        num_h_patches, num_w_patches = num_h_patches + 0.1, num_w_patches + 0.1
+        sqrt_num_positions = math.sqrt(num_positions)
+        patch_pos_embed = patch_pos_embed.reshape(1, int(sqrt_num_positions), int(sqrt_num_positions), embed_dim)
+        patch_pos_embed = patch_pos_embed.permute(0, 3, 1, 2)
+        fp32_upcasting = patch_pos_embed.dtype == torch.bfloat16
+        if fp32_upcasting:
+            logger.warning_once(
+                "Upcasting patch_pos_embed to fp32 for interpolation since upsample_bicubic2d_out_frame in nn.functional.interpolate"
+                "is not implemented for 'BFloat16'. This will result in a slight overhead"
```
Comment (ArthurZucker) [inline src/transformers/models/idefics/vision.py:119]:
```suggestion
                "Upcasting patch_pos_embed to fp32 for interpolation since `upsample_bicubic2d_out_frame` in nn.functional.interpolate"
                "is not implemented for 'torch.bfloat16' dtype. This will result in a slight overhead"
```

File: src/transformers/models/idefics/vision.py [hunk 10]
```diff
@@ -87,15 +85,79 @@ def __init__(self, config: IdeficsVisionConfig):
...
+        patch_pos_embed = patch_pos_embed.permute(0, 2, 3, 1).view(1, -1, embed_dim)
+        return torch.cat((class_pos_embed.unsqueeze(0), patch_pos_embed), dim=1)
+
+    def forward(self, pixel_values: torch.FloatTensor, interpolate_pos_encoding: bool = False) -> torch.Tensor:
+        batch_size, num_channels, height, width = pixel_values.shape
+        if not interpolate_pos_encoding:
+            if height != self.image_size or width != self.image_size:
+                raise ValueError(
+                    f"Input image size ({height}*{width}) doesn't match model"
+                    f" ({self.image_size}*{self.image_size})."
```
Comment (ArthurZucker) [inline src/transformers/models/idefics/vision.py:144]:
```suggestion
                    f"Input image size ({height}*{width}) doesn't match model"
                    f" ({self.image_size}*{self.image_size}). You should try to set `interpolate_pos_encoding=True`"
```

File: tests/models/idefics/test_modeling_idefics.py [hunk 11]
```diff
@@ -268,10 +293,34 @@ def setUp(self):
...
+        )
+        self.model_tester.create_and_check_model(*config_and_inputs)
+
+    def test_model_multiple_images(self):
+        config_and_inputs = self.model_tester.prepare_config_and_inputs(
+            num_images=2, interpolate_pos_encoding=False, image_expansion=0
+        )
         self.model_tester.create_and_check_model(*config_and_inputs)
 
+    def test_model_with_image_pos_embeddings_interpolation(self):
```
Comment (ArthurZucker) [inline tests/models/idefics/test_modeling_idefics.py:308]: interpolate is never tested with 2 images is this expected to not work?

**Review by ArthurZucker — APPROVED** [review]
Looks good to me left a few nits

File: tests/models/idefics/test_modeling_idefics.py (hunk 11 above)
Comment (leot13) [reply tests/models/idefics/test_modeling_idefics.py]: No, the opposite. I added the multiple image tests for interpolation and generation
